### ISSUE TITLE
Request: UI improvements to dataset browser

### ISSUE THREAD

**Issue by marshall007** [solvebio/solvebio#2]
Here's a few suggestions for improvements to the dataset browser UI:

- Tree view as an alternative to existing table view (similar to the [JSON formatter](https://chrome.google.com/webstore/detail/json-formatter/bcjindcccaagfpapjjmafapmmgkkhgoa?hl=en) Chrome extension)
- Toggle which columns are visible in table view
  - Automatically hide columns that are empty for all rows?
- "Fullscreen" view
  - Quickly switch to other datasets from dropdowns without having to leave the browse mode
  - Allow viewing more than 25 results at a time
- Select and copy row(s) as JSON
- Improve scroll performance (lag) of table view

**Comment by davecap** [comment]
Thanks for the suggestions Marshall!

Our top priority right now is improving the performance of the table view. We should have that done by next week. Showing more than 25 results at a time is possible, but is a significant factor in the performance of the table.

Next up would be column toggling. Instead of auto-hiding empty columns, we will probably set default visible columns per dataset to avoid the empty cell issue.

Copying rows is a nice one too. We are working on a bulk export feature in which it will be possible to export the first N rows of a result-set.

The full screen mode with dataset switching is a good idea, and I'll discuss it with the team to figure out where that fits on our roadmap.

**Comment by marshall007** [comment]
Any thoughts on the JSON tree view as an alternative? We'd probably use that more than the current table view and having it would make the other things less important.

> Copying rows is a nice one too. We are working on a bulk export feature in which it will be possible to export the first N rows of a result-set.

I said rows here, but in addition to exporting the idea is just to make it easier to copy out any text value from the table. It's currently difficult to select the text in a field especially if it's an array or spans multiple lines.

**Comment by davecap** [comment]
Yes, we've got a single record view (basically a JSON tree) up next in our backlog, so expect to see it soon on the site!

From what you're saying, would it be useful enough to have the single record (tree-like) view for copying out values from records?

**Comment by marshall007** [comment]
@davecap yea, I think having the tree view would pretty much solve that piece. I would prefer to switch between table and tree view for all records rather than opening up a tree view for a single record, though. I was thinking it would look something like this:

![screen shot 2015-08-06 at 6 10 26 pm](https://cloud.githubusercontent.com/assets/877936/9125540/945b82da-3c67-11e5-9038-973645356f40.png)

**Comment by davecap** [comment]
We pushed some performance changes to the data table yesterday which should solve the lag problem. Try it out and let us know what you think! The record/card/tree view is coming up next, we've made some progress on it and hope to deploy it early next week.

**Comment by tayloramurphy** [comment]
The performance seems much better. Personally, I like that it jumps to each column instead of trying to smooth it out. 

I'll give another vote for being able to add/remove columns as needed. It's nice that the list is on the Documentation tab, but being able to toggle them from a dropdown list would be handy. But there might be a better UI for that.

Nice work y'all!

**Comment by davecap** [comment]
Thanks! The column selector is definitely high on the roadmap, we'll keep you posted.

**Comment by davecap** [comment]
@marshall007 we've added a "view mode" to the dataset browser. You can switch between table, JSON, and record views. The JSON and record views have a copy button too. Give it a try and let us know what you think.

**Comment by tayloramurphy** [comment]
The JSON and record views look awesome! The Copy JSON feature is a nice touch as well. Great work on this, it's really nice.

**Comment by davecap** [comment]
Glad you like it!

**Comment by marshall007** [comment]
@davecap this is almost exactly what I had in mind, looks great! I quite like the dedicated `Copy JSON` button as well!

**Comment by davecap** [comment]
Awesome! We're still working on a few improvements. If you notice anything weird let us know!
